server/api/strategicgroups/index.ts

```diff
@@ -0,0 +1,262 @@
+    user.contribYear = parseInt(contribYear.numcontribs, 10);
+  }
+
+  return { group, projects, users };
+}
+
+export async function getStrategicProject(req, id) {
+  const project = (await dbprojects.searchProjectById(id))[0]; // returns one project as a list
+  const groups = await dbgroups.getGroupsByProjectId([id]);
+  const groupIds = (await dbgroups.searchGroupIdsByProjectId(id)).groups;
+  const users = (await dbusers.getUsernamesByGroup(groupIds)).names;
+  const projId = [project.project_id];
+  const userList = [];
+
+  for (const group of groups) {
+    const usernames = (await dbusers.getUsernamesByGroup([
+      group.group_id.toString(),
+    ])).names;
+    const contribWeek = await dbcontributions.getLastWeekCount(
+      projId,
+      usernames
+    );
+    const contribMTD = await dbcontributions.getMTDCount(projId, usernames);
+    const contribMonth = await dbcontributions.getLastMonthCount(
+      projId,
+      usernames
+    );
+    const contribYear = await dbcontributions.getYTDCount(projId, usernames);
+
+    group.contribWeek = parseInt(contribWeek.numcontribs, 10);
+    group.contribMTD = parseInt(contribMTD.numcontribs, 10);
+    group.contribMonth = parseInt(contribMonth.numcontribs, 10);
+    group.contribYear = parseInt(contribYear.numcontribs, 10);
+  }
+
+  for (const user of users) {
+    const contribWeek = await dbcontributions.getLastWeekCount(projId, [user]);
+    const contribMTD = await dbcontributions.getMTDCount(projId, [user]);
+    const contribMonth = await dbcontributions.getLastMonthCount(projId, [
+      user,
+    ]);
+    const contribYear = await dbcontributions.getYTDCount(projId, [user]);
+
+    const data = {
+      amazon_alias: user,
+      contribWeek: parseInt(contribWeek.numcontribs, 10),
+      contribMTD: parseInt(contribMTD.numcontribs, 10),
+      contribMonth: parseInt(contribMonth.numcontribs, 10),
+      contribYear: parseInt(contribYear.numcontribs, 10),
+    };
+    userList.push(data);
+  }
+  return { project, groups, users: userList };
+}
+
+export async function listProjects(req) {
+  return { projectList: await dbprojects.getProjectsByGroup(req.body.id) };
+}
+
+export async function listUsers(req) {
+  return { users: await dbusers.listAllUsers() };
+}
+
+export async function addNewGroup(req, body) {
+  const groupId = await dbgroups.addNewGroup(
+    body.groupName,
+    body.sponsorName,
+    body.goals,
+    body.projects
+  );
+  for (const user of body.users) {
+    const id = groupId.group_id;
+    const date = new Date().toISOString().substring(0, 10);
+    const group = `{"${id}":"${date}"}`;
+
+    await dbusers.addGroupsToUser(user, group);
+  }
+  return { groupId };
+}
+
+export async function addNewUser(req, body) {
+  let groups = '{';
+  if (body.groups.length) {
+    const date = new Date().toISOString().substring(0, 10);
+    for (const group of body.groups) {
+      groups += `"${group}":"${date}",`;
```

Comment:
There's a *potential* SQL injection problem here, though it's somewhat constrained to the column where you're storing groups in this call. Consider constructing a native object for `groups` (the pg driver can figure it out) instead of stringing this together. (Check the other locations where you're building similar strings too.)

Reply:
I didn't realize the pg driver would interpret the object! I have now changed it to form a json object so that it's safer.